Godot 4.0 is very slow to start on macOS compared to 3.x

### Godot version

4.0.alpha10.official

### System information

macOS 12.4 (Monterey), Vulkan, Intel Iris Pro 5200

### Issue description

Latest release of Godot 4.0 (alpha 10) seems to be super slow on macOS 12.4 (Monterey), at least when tested on my MacBook Pro setup (meanwhile, Godot 3.4.4 works without any noticeable performance issues). Here are some benchmarks:

- time to start Godot (to open project list): **33 seconds**
- time to open new project: 35 seconds
- time to start an empty 2D scene app: 33 seconds

On the other hand, editor itself is generally responsive when interacting with its UI (creating new nodes etc.).

Among important details, here are some errors that are raised during the 2D application start (empty scene):
```
E 0:00:00:0276   _debug_messenger_callback:  - Message Id Number: 0 | Message Id Name: 
	VK_ERROR_INITIALIZATION_FAILED: Render pipeline compile failed (Error code 2):
Compiler encountered an internal error.
	Objects - 1
		Object[0] - VK_OBJECT_TYPE_PIPELINE, Handle 140389825172992
  <C++ Source>   drivers/vulkan/vulkan_context.cpp:159 @ _debug_messenger_callback()

E 0:00:00:0276   render_pipeline_create: vkCreateGraphicsPipelines failed with error -3 for shader 'ClusterRenderShaderRD:0'.
  <C++ Error>    Condition "err" is true. Returning: RID()
  <C++ Source>   drivers/vulkan/rendering_device_vulkan.cpp:6614 @ render_pipeline_create()

E 0:00:30:0395   _debug_messenger_callback:  - Message Id Number: 0 | Message Id Name: 
	VK_ERROR_INITIALIZATION_FAILED: Render pipeline compile failed (Error code 2):
Compiler encountered an internal error.
	Objects - 1
		Object[0] - VK_OBJECT_TYPE_PIPELINE, Handle 140389832876544
  <C++ Source>   drivers/vulkan/vulkan_context.cpp:159 @ _debug_messenger_callback()

E 0:00:30:0395   render_pipeline_create: vkCreateGraphicsPipelines failed with error -3 for shader 'ClusterRenderShaderRD:0'.
  <C++ Error>    Condition "err" is true. Returning: RID()
  <C++ Source>   drivers/vulkan/rendering_device_vulkan.cpp:6614 @ render_pipeline_create()
```

Also, here's the editor output after Godot start, for the record:
```
--- Debugging process started ---
Godot Engine v4.0.alpha10.official.4bbe7f0b9 - https://godotengine.org
Vulkan API 1.1.198 - Using Vulkan Device #0: Intel - Intel Iris Pro Graphics
 
Registered camera FaceTime HD Camera with id 1 position 0 at index 0
--- Debugging process stopped ---
```

It seems that Vulkan support is not an issue on my MacBook Pro. I've just installed Vulkan SDK 1.3.216.0 for macOS from [LunarG website](https://vulkan.lunarg.com/sdk/home#mac) and their `vkcube.app` demo works without any issues (thanks to [MoltenVK](https://github.com/KhronosGroup/MoltenVK) underneath, of course).

I can provide any additional information or testing if that would help.

### System details
- laptop: MacBook Pro (Retina, 15-inch, Mid 2015)
- CPU: Intel Core i7-4870HQ 2.5 GHz (Haswell)
- OS: macOS Monterey 12.4 (build [21F79](https://en.wikipedia.org/wiki/MacOS_Monterey#Release_history))

### GPU details (from `system_profiler`)
```
Chipset Model: Intel Iris Pro
Type: GPU
Bus: Built-In
VRAM (Dynamic, Max): 1536 MB
Vendor: Intel
Device ID: 0x0d26
Revision ID: 0x0008
Metal Family: Supported, Metal GPUFamily macOS 1
```

And by the way, thank you all for Godot! 🎉 It's absolutely amazing that a game engine like this is available out there.

### Steps to reproduce

No extra steps are needed to reproduce the issue. It seems that any MacBook with a similar spec would be prone to this problem.

### Minimal reproduction project

_No response_

I have this problem as well and looked into it a bit. I don't have a solution but my findings so far is that one call to vkCreateGraphicsPipelines (for the shader ClusterRenderShaderRD:0) takes 30s. What happens is that the separate MTLCompilerService process crashes, and the caller times out & continues after 10 seconds. (It does 3 retries however, hence a 30s total delay). So it doesn't seem to be a shader cache performance issue per se like @Calinou investigated, but a bug, possibly in the MoltenVK/Metal backend.

As for why it's crashing is a bit more unclear. The callstack of the crashed thread in MTLCompilerService is:
```Thread 2 Crashed:
0   libsystem_kernel.dylib        	    0x7ff819438dba __abort_with_payload + 10
1   libsystem_kernel.dylib        	    0x7ff81943a877 abort_with_payload_wrapper_internal + 80
2   libsystem_kernel.dylib        	    0x7ff81943a827 abort_with_reason + 19
3   MTLCompiler                   	    0x7ffa32221d4f fatalErrorHandler(void*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, bool) + 635
4   libLLVM.dylib                 	    0x7ffa31e86986 llvm::report_fatal_error(llvm::Twine const&, bool) + 323
5   libLLVM.dylib                 	    0x7ffa31e869c0 llvm::report_fatal_error(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, bool) + 29
6   libIGIL-Metal.dylib           	    0x7ffa295242dd IGILErrorMessageHandlers::reportUnsupportedFunctionReferenceError(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) + 42
7   libIGIL-Metal.dylib           	    0x7ffa2954532f (anonymous namespace)::BiFInliner::runOnModule(llvm::Module&) + 3877
8   libLLVM.dylib                 	    0x7ffa31af3c15 llvm::legacy::PassManagerImpl::run(llvm::Module&) + 559
9   libMTLIntelCompilerPlugin.dylib	       0x10291acb9 MTLIntelCompiler::generateIGIL(llvm::Module*, MTLIntelFunctionType, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&, bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&) + 669
10  libMTLIntelCompilerPlugin.dylib	       0x10290fa3b MTLCompilerBuildRequestWithOptions + 523
11  MTLCompiler                   	    0x7ffa3221276a MTLCompilerPluginInterface::compilerBuildRequest(bool, unsigned int, void const*, unsigned long, unsigned int, void const*, BackendCompilationOutput&) + 276
12  MTLCompiler                   	    0x7ffa3220cd11 MTLCompilerObject::backendCompileModule(MTLCompilerObject::BinaryRequestData&, BackendCompilationOutput&, unsigned long long, std::__1::vector<CompileTimeData, std::__1::allocator<CompileTimeData> >&) + 159
13  MTLCompiler                   	    0x7ffa322129f7 MTLCompilerObject::backendCompileExecutableRequest(MTLCompilerObject::BinaryRequestData&) + 517
14  MTLCompiler                   	    0x7ffa32221658 MTLCompilerObject::buildRequest(unsigned int, unsigned int, void const*, unsigned long, void (unsigned int, void const*, unsigned long, char const*) block_pointer) + 258
15  MTLCompiler                   	    0x7ffa321f822d split_stack_call + 13
16  MTLCompiler                   	    0x7ffa32212cef MTLCodeGenServiceBuildRequest + 265
17  MTLCompilerService            	       0x1026d57e3 invocation function for block in MTLCompilerServiceHandleEvent(NSObject<OS_xpc_object>*) + 793
18  libxpc.dylib                  	    0x7ff819191b70 _xpc_connection_call_event_handler + 56
19  libxpc.dylib                  	    0x7ff819190956 _xpc_connection_mach_event + 1413
20  libdispatch.dylib             	    0x7ff81929b3b1 _dispatch_client_callout4 + 9
21  libdispatch.dylib             	    0x7ff8192b4041 _dispatch_mach_msg_invoke + 445
22  libdispatch.dylib             	    0x7ff8192a11cd _dispatch_lane_serial_drain + 342
23  libdispatch.dylib             	    0x7ff8192b4b77 _dispatch_mach_invoke + 484
24  libdispatch.dylib             	    0x7ff8192a11cd _dispatch_lane_serial_drain + 342
25  libdispatch.dylib             	    0x7ff8192a1dfd _dispatch_lane_invoke + 366
26  libdispatch.dylib             	    0x7ff8192abeee _dispatch_workloop_worker_thread + 753
27  libsystem_pthread.dylib       	    0x7ff81944efd0 _pthread_wqthread + 326
28  libsystem_pthread.dylib       	    0x7ff81944df57 start_wqthread + 15
```

It seems to me like it encounters an error, but fails to properly report it and crashes instead. `reportUnsupportedFunctionReferenceError` is seems like the biggest hint to to the cause to me. I have not been able to attach to the process with a debugger to try to extract more info. (The process is a bit transient.)

While compiling it also emits ``[MTLCompiler] Error: undefined reference to `air_simd_is_helper_thread()'`` to the system log. This happens more times than for the failing call though so I'm not sure it's related.

I don't know if there are any more Vulkan/Metal logs to check, but let me know if there's some place I should have a peek.

Ok so I did a bit of a deep dive on this one, and basically yes the Metal compiler crashes and `[MTLCompiler] Error: undefined reference to `air_simd_is_helper_thread()` is the closest to a detailed error there is. Unfortunately this is in internal proprietary code and not much to work with so that's a bit of a dead end.

However, with a bit of testing I figured out that what actually _triggers_ this error is the usage of `gl_HelperInvocation` in cluster_render.glsl:

https://github.com/godotengine/godot/blob/c660cc4adc7032c11f74dfce5fcb7b5a02f6d097/servers/rendering/renderer_rd/shaders/cluster_render.glsl#L164

In my case it's taking the non-`USE_SUBGROUPS`path and removing those two conditionals (i.e. running the `atomicOr` always) fixes the compiler failure and Godot starts fast.

I'm not sure making a pull request of that makes sense because those conditionals are obviously there for a reason, and I'm not well versed enough in the shaders to know what the best alternative is. (As I understand it they're there to avoid doing unnecessary calculations when the shader is run in a helper mode that doesn't actually output anything, so removing it likely doesn't cause any errors but may have performance effects. If I read the docs correctly atomicOr doesn't actually do anything when in a helper invocation though so possibly it's just basically a free NOP anyway, but someone who actually knows what they're talking about should chime in.)

As an extra reference, this specific code is actually discussed in a separate but related discussion about subgroups + helper invocations in a Khronos GLSL issue, mentioning @reduz : https://github.com/KhronosGroup/GLSL/issues/35

I'd be happy to test alternative fixes on this computer where it normally fails, just let me know.

--- a/servers/rendering/renderer_rd/shaders/cluster_render.glsl
+++ b/servers/rendering/renderer_rd/shaders/cluster_render.glsl
@@ -141,7 +141,11 @@ void main() {
 		}
 	}
 #else
-	if (!gl_HelperInvocation) {
+// MoltenVK/Metal fails to compile shaders using gl_HelperInvocation for some GPUs
+#ifndef MOLTENVK_USED
+	if (!gl_HelperInvocation)
+#endif
+	{
 		atomicOr(cluster_render.data[usage_write_offset], usage_write_bit);
 	}
 #endif
@@ -161,7 +165,11 @@ void main() {
 		}
 	}
 #else
-	if (!gl_HelperInvocation) {
+// MoltenVK/Metal fails to compile shaders using gl_HelperInvocation for some GPUs
+#ifndef MOLTENVK_USED
+	if (!gl_HelperInvocation)
+#endif
+	{
 		atomicOr(cluster_render.data[z_write_offset], z_write_bit);
 	}
 #endif

